fix: platforms, architectures and dpkg_status after #115

jjmaestro opened the pull request:
After landing #115 we need `platforms` to be non-dev. This would have been caught in the e2e test but since #115 didn't make  changes to the e2e test, it went unnoticed.

To prevent this, I've also copied the `examples/debian_snapshot` to the `e2e` test. IMHO this helps catch bugs that affect third-party repos using `rules_distroless` as a dependency, such as the `platforms` dependency.

I've also verified that we do need the `select` in the `architecture`, as mentioned in my [code review](https://github.com/GoogleContainerTools/rules_distroless/pull/115#discussion_r1855644693). Otherwise, the `Config` in the manifest is always set to `arm64` regardless of the architecture that we are building.

I've also fixed the unmapped Debian architectures (e.g. "all"). Basically, `_ARCHITECTURE_MAP[arch] or arch` fails when a Debian arch is not in the map. It should be `_ARCHITECTURE_MAP.get(arch, arch)` to default to `arch` when it's not in the map. To exercise this and ensure it always works I've:

* removed "arm64" since it's a valid CPU platform already (and without this fix it would have already failed in all the examples and the e2e test).
* added `armhf` mapping to `armv7e-mf`
* changed Debian's `ppc64el` mapping to the exact matching platforms CPU (`ppc64le`)
* added a test to exercise arch `all` (which is both an arch that's not in the map while being "a special Debian arch" and also a valid platforms CPU constraint).
* Moved architecture doc links next to `_ARCHITECTURE_MAP` and added a bunch more links to the Debian wikis.

Finally, I've fixed `_PACKAGES`, `:packages` and `:dpkg_status` to be different per-architecture, because there are cases where the packages differ between architectures. E.g. currently in the `examples/ubuntu_snapshot`, `coreutils` depends on `libssl3` on `amd64` but there's no such dependency on `arm64`. Thus, the list of packages and `dpkg_status` has to be different.

thesayyn commented:
I am little at unease about the symlinks, this will most likely break BCR, because of https://github.com/GoogleContainerTools/rules_distroless/blob/7b342e2803ea0a98a8464c786d5d2b710a5c1a14/.gitattributes#L10 and how BCR uses release archives (created via git archive with examples folder excluded) to run presubmit.

Can we just improve e2e so it catches this case?

jjmaestro commented:
Ah, I didn't know about BCR! Sure, I'll copy over the test and will make a note in it to keep it in sync.

jjmaestro commented:
@thesayyn Added another fix I caught while adding one of the tests that I wrote for #100 :)  Let me know if you like the PR and let the workflows run when you can (I've manually tested the PR so it should hopefully run green). Thanks!

jjmaestro commented:
@thesayyn Added yet-another fix, the "new" `dpk_status` from #115 would fail if there were different packages installed per-architecture. This is actually the case in at least one of the examples in the repo. See b0cc2da for more details.

thesayyn commented on apt/private/deb_translate_lock.bzl:
```diff
@@ -44,22 +44,30 @@ load("@rules_distroless//apt:defs.bzl", "dpkg_status")
+#     * https://github.com/bazelbuild/platforms/blob/main/cpu/BUILD
 _ARCHITECTURE_MAP = {{
     "amd64": "x86_64",
-    "arm64": "aarch64",
```
this is a great catch, should we add all possible combination even if they are identical just for completeness? 
eg "arm64": "arm64".

jjmaestro replied:
Ah, I see! That's probably better than fall-though, set all the architectures that should be there and fail if not in the dict.

OK, will add them to the map and will re-word the commit!

thesayyn commented on examples/debian_snapshot_all_arch/bullseye_all.yaml:
```diff
@@ -0,0 +1,20 @@
+# Packages for examples/debian_snapshot_all_arch.
```
Can you move this test into [apt/tests/resolution](https://github.com/GoogleContainerTools/rules_distroless/tree/main/apt/tests/resolution) folder? trying to keep examples helpful

thesayyn commented on MODULE.bazel:
```diff
@@ -11,6 +11,7 @@ bazel_dep(name = "protobuf", version = "29.0-rc2.bcr.1")
 bazel_dep(name = "bazel_skylib", version = "1.5.0")
 bazel_dep(name = "aspect_bazel_lib", version = "2.9.4")
 bazel_dep(name = "rules_java", version = "8.3.2")
+bazel_dep(name = "platforms", version = "0.0.10")
```
This is already fixed on main, you can remove it

thesayyn commented:
> would fail if there were different packages installed per-architecture

Yup, this is a great catch and paves the way for per architecture debian repositories and packages https://github.com/GoogleContainerTools/rules_distroless/issues/56

jjmaestro commented:
Oh wow, I just fixed the PR and I was checking "Compare" and the rebase still shows!!? even when it's rebased over `main`! 😭 I was expecting that GH would separate rebasing over `main` VS e.g. other branches and would NOT include those as "changes"... at least provide a way to see "all changes" and "changes VS last state of the PR".

@thesayyn  this, again, will make it difficult to review, sorry 😞  It'll be easier if you check the final state of the commits, but basically, I've done as you requested in your comments.

jjmaestro commented:
OK, I've produced such a diff myself with:

```shell
git co -b fix-pr-115-OLD b0cc2da
git diff fix-pr-115-OLD fix-pr-115 -- . ':(exclude)*.lock.json'
```

But GH doesn't let me attach it as a file to the comment!?? srsly... it seems to only allow PDF, images, movies, etc.

Anyway, here you go, in case it makes the review easier:

```diff
diff --git a/MODULE.bazel b/MODULE.bazel
index 35dfbc9..7c598e5 100644
--- a/MODULE.bazel
+++ b/MODULE.bazel
@@ -13,7 +13,6 @@ bazel_dep(name = "protobuf", version = "29.0-rc2.bcr.1")
 bazel_dep(name = "bazel_skylib", version = "1.5.0")
 bazel_dep(name = "aspect_bazel_lib", version = "2.9.4")
 bazel_dep(name = "rules_java", version = "8.3.2")
-bazel_dep(name = "platforms", version = "0.0.10")
 
 bazel_lib_toolchains = use_extension("@aspect_bazel_lib//lib:extensions.bzl", "toolchains")
 use_repo(bazel_lib_toolchains, "zstd_toolchains")
@@ -69,13 +68,7 @@ apt.install(
 )
 apt.install(
     name = "bullseye_nolock",
-    manifest = "//examples/debian_snapshot:bullseye_nolock.yaml",
-    nolock = True,
-)
-apt.install(
-    name = "bullseye_all",
-    lock = "//examples/debian_snapshot_all_arch:bullseye_all.lock.json",
-    manifest = "//examples/debian_snapshot_all_arch:bullseye_all.yaml",
+    manifest = "//examples/debian_snapshot:bullseye.yaml",
     nolock = True,
 )
 apt.install(
@@ -88,4 +81,9 @@ apt.install(
     manifest = "apt/tests/resolution/security.yaml",
     nolock = True,
 )
-use_repo(apt, "bullseye", "bullseye_all", "bullseye_nolock", "noble", "resolution_test", "resolution_test_resolve")
+apt.install(
+    name = "arch_all_test",
+    manifest = "apt/tests/resolution/arch_all.yaml",
+    nolock = True,
+)
+use_repo(apt, "bullseye", "bullseye_nolock", "noble", "resolution_test", "resolution_test_resolve", "arch_all_test", "arch_all_test_resolve")
diff --git a/apt/private/deb_translate_lock.bzl b/apt/private/deb_translate_lock.bzl
index 7a924bf..15c5b31 100644
--- a/apt/private/deb_translate_lock.bzl
+++ b/apt/private/deb_translate_lock.bzl
@@ -56,9 +56,13 @@ exports_files(['packages.bzl'])
 #     * https://github.com/bazelbuild/platforms/blob/main/cpu/BUILD
 _ARCHITECTURE_MAP = {{
     "amd64": "x86_64",
-    "armhf": "armv7e-mf",
-    "mips64el": "mips64",
+    "arm64": "arm64",
     "ppc64el": "ppc64le",
+    "mips64el": "mips64",
+    "s390x": "s390x",
+    "i386": "x86_32",
+    "armhf": "armv7e-mf",
+    "all": "all",
 }}
 
 _ARCHITECTURES = {architectures}
@@ -68,7 +72,7 @@ _ARCHITECTURES = {architectures}
     name = os + "_" + arch,
     constraint_values = [
        "@platforms//os:" + os,
-       "@platforms//cpu:" + _ARCHITECTURE_MAP.get(arch, arch),
+       "@platforms//cpu:" + _ARCHITECTURE_MAP[arch],
     ],
   )
   for os in ["linux"]
diff --git a/apt/tests/resolution/BUILD.bazel b/apt/tests/resolution/BUILD.bazel
index 2a64bb3..d799b43 100644
--- a/apt/tests/resolution/BUILD.bazel
+++ b/apt/tests/resolution/BUILD.bazel
@@ -15,3 +15,33 @@ assert_contains(
     actual = ":pick_libuuid_version",
     expected = "2.38.1-5+deb12u1",
 )
+
+jq(
+    name = "pick_quake_arch",
+    srcs = [
+        "@arch_all_test_resolve//:lockfile",
+    ],
+    args = ["-rj"],
+    filter = '.packages | map(select(.name == "quake")) | .[0].arch',
+)
+
+assert_contains(
+    name = "test_quake_arch",
+    actual = ":pick_quake_arch",
+    expected = "all",
+)
+
+jq(
+    name = "pick_quake_version",
+    srcs = [
+        "@arch_all_test_resolve//:lockfile",
+    ],
+    args = ["-rj"],
+    filter = '.packages | map(select(.name == "quake")) | .[0].version',
+)
+
+assert_contains(
+    name = "test_quake_version",
+    actual = ":pick_quake_version",
+    expected = "73",
+)
diff --git a/apt/tests/resolution/arch_all.yaml b/apt/tests/resolution/arch_all.yaml
new file mode 100644
index 0000000..537c993
--- /dev/null
+++ b/apt/tests/resolution/arch_all.yaml
@@ -0,0 +1,11 @@
+version: 1
+
+sources:
+  - channel: bookworm main contrib
+    url: https://snapshot-cloudflare.debian.org/archive/debian/20240401T030239Z
+
+archs:
+  - all
+
+packages:
+  - quake
diff --git a/e2e/smoke/BUILD b/e2e/smoke/BUILD
index 88b98ba..73e6323 100644
--- a/e2e/smoke/BUILD
+++ b/e2e/smoke/BUILD
@@ -12,7 +12,7 @@ load("@aspect_bazel_lib//lib:tar.bzl", "tar")
 load("@aspect_bazel_lib//lib:transitions.bzl", "platform_transition_filegroup")
 load("@container_structure_test//:defs.bzl", "container_structure_test")
 load("@rules_distroless//distroless:defs.bzl", "cacerts", "group", "passwd")
-load("@rules_oci//oci:defs.bzl", "oci_image", "oci_load")
+load("@rules_oci//oci:defs.bzl", "oci_image")
 
 passwd(
     name = "passwd",
@@ -75,9 +75,10 @@ oci_image(
         "SSL_CERT_FILE": "/etc/ssl/certs/ca-certificates.crt",
     },
     os = "linux",
+    tags = ["manual"],
     tars = [
         # This target contains all the installed packages.
-        "@bullseye//:bullseye",
+        "@bullseye//:flat",
         ":sh",
         ":passwd",
         ":group",
@@ -110,13 +111,13 @@ platform_transition_filegroup(
     }),
 )
 
-oci_load(
-    name = "tarball",
-    image = ":image_platform",
-    repo_tags = [
-        "distroless/test:latest",
-    ],
-)
+# oci_load(
+#     name = "tarball",
+#     image = ":image_platform",
+#     repo_tags = [
+#         "distroless/test:latest",
+#     ],
+# )
 
 container_structure_test(
     name = "test",
diff --git a/e2e/smoke/bullseye.yaml b/e2e/smoke/bullseye.yaml
index 2c3052c..7bac5e7 100644
--- a/e2e/smoke/bullseye.yaml
+++ b/e2e/smoke/bullseye.yaml
@@ -11,12 +11,11 @@ version: 1
 
 sources:
   - channel: bullseye main contrib
-    url: https://snapshot.debian.org/archive/debian/20240210T223313Z
+    url: https://snapshot-cloudflare.debian.org/archive/debian/20240210T223313Z
   - channel: bullseye-security main
-    url: https://snapshot.debian.org/archive/debian-security/20240210T223313Z
+    url: https://snapshot-cloudflare.debian.org/archive/debian-security/20240210T223313Z
   - channel: bullseye-updates main
-    url: https://snapshot.debian.org/archive/debian/20240210T223313Z/
-  # This channel contains an uncompressed Packages file 
+    url: https://snapshot-cloudflare.debian.org/archive/debian/20240210T223313Z/
   - channel: cloud-sdk main
     url: https://packages.cloud.google.com/apt
 
@@ -28,15 +27,13 @@ packages:
   - "ncurses-base"
   - "libncurses6"
   - "tzdata"
-  - "bash"
-  - "coreutils" # for commands like `ls`
-  # for apt list --installed
+  - "coreutils"
   - "dpkg"
   - "apt"
   - "perl"
-  # test cacerts() compatibility
   - "ca-certificates"
-  # test multiple components: contrib
   - "nvidia-kernel-common"
-  # test uncompressed Package
-  - google-cloud-cli
+  - "bash"
+  - "nginx-full"
+  - "nginx-core"
+  - "google-cloud-cli"
diff --git a/examples/debian_snapshot/bullseye.yaml b/examples/debian_snapshot/bullseye.yaml
index fcbfaf0..7bac5e7 100644
--- a/examples/debian_snapshot/bullseye.yaml
+++ b/examples/debian_snapshot/bullseye.yaml
@@ -11,11 +11,13 @@ version: 1
 
 sources:
   - channel: bullseye main contrib
-    url: https://snapshot.debian.org/archive/debian/20240210T223313Z
+    url: https://snapshot-cloudflare.debian.org/archive/debian/20240210T223313Z
   - channel: bullseye-security main
-    url: https://snapshot.debian.org/archive/debian-security/20240210T223313Z
+    url: https://snapshot-cloudflare.debian.org/archive/debian-security/20240210T223313Z
   - channel: bullseye-updates main
-    url: https://snapshot.debian.org/archive/debian/20240210T223313Z/
+    url: https://snapshot-cloudflare.debian.org/archive/debian/20240210T223313Z/
+  - channel: cloud-sdk main
+    url: https://packages.cloud.google.com/apt
 
 archs:
   - "amd64"
@@ -34,3 +36,4 @@ packages:
   - "bash"
   - "nginx-full"
   - "nginx-core"
+  - "google-cloud-cli"
diff --git a/examples/debian_snapshot/bullseye_nolock.yaml b/examples/debian_snapshot/bullseye_nolock.yaml
deleted file mode 100644
index d5e9cc9..0000000
--- a/examples/debian_snapshot/bullseye_nolock.yaml
+++ /dev/null
@@ -1,12 +0,0 @@
-version: 1
-
-sources:
-  - channel: bullseye main
-    url: https://snapshot.debian.org/archive/debian/20240210T223313Z
-
-archs:
-  - "amd64"
-  - "arm64"
-
-packages:
-  - "bash"
diff --git a/examples/debian_snapshot_all_arch/BUILD.bazel b/examples/debian_snapshot_all_arch/BUILD.bazel
deleted file mode 100644
index 4a27aae..0000000
--- a/examples/debian_snapshot_all_arch/BUILD.bazel
+++ /dev/null
@@ -1,32 +0,0 @@
-load("@aspect_bazel_lib//lib:jq.bzl", "jq")
-load("@aspect_bazel_lib//lib:testing.bzl", "assert_contains")
-
-jq(
-    name = "pick_quake_arch",
-    srcs = [
-        "@bullseye_all_resolve//:lockfile",
-    ],
-    args = ["-rj"],
-    filter = '.packages | map(select(.name == "quake")) | .[0].arch',
-)
-
-assert_contains(
-    name = "test_quake_arch",
-    actual = ":pick_quake_arch",
-    expected = "all",
-)
-
-jq(
-    name = "pick_quake_version",
-    srcs = [
-        "@bullseye_all_resolve//:lockfile",
-    ],
-    args = ["-rj"],
-    filter = '.packages | map(select(.name == "quake")) | .[0].version',
-)
-
-assert_contains(
-    name = "test_quake_version",
-    actual = ":pick_quake_version",
-    expected = "67",
-)
diff --git a/examples/debian_snapshot_all_arch/bullseye_all.yaml b/examples/debian_snapshot_all_arch/bullseye_all.yaml
deleted file mode 100644
index 7f92d58..0000000
--- a/examples/debian_snapshot_all_arch/bullseye_all.yaml
+++ /dev/null
@@ -1,20 +0,0 @@
-# Packages for examples/debian_snapshot_all_arch.
-#
-#  Anytime this file is changed, the lockfile needs to be regenerated.
-#
-#  To generate the bullseye_all.lock.json run the following command
-#
-#     bazel run @bullseye_all//:lock
-#
-# See debian_package_index at WORKSPACE.bazel
-version: 1
-
-sources:
-  - channel: bullseye contrib
-    url: https://snapshot-cloudflare.debian.org/archive/debian/20240210T223313Z
-
-archs:
-  - all
-
-packages:
-  - quake
```

thesayyn commented:
You have some buildifier errors.

thesayyn commented:
https://github.com/GoogleContainerTools/rules_distroless/actions/runs/12284130637/job/34279199329?pr=120#step:3:40

jjmaestro commented:
@thesayyn I've created #135 from `main` that fixes some buildifier stuff that got through the cracks (somehow didn't run `pre-commit run` I think?). I'll rebase on that one and see if there's anything left here as well.

jjmaestro commented:
OK, this is with the rebase, now fixing the broken buildifiers left here...

thesayyn commented:
> @thesayyn I've created #135 from `main` that fixes some buildifier stuff that got through the cracks (somehow didn't run `pre-commit run` I think?). I'll rebase on that one and see if there's anything left here as well.

It's just Readme files over there, which is not related to your changes, you have some unused loads.

thesayyn commented:
This made macos unhappy.

thesayyn commented:
https://github.com/GoogleContainerTools/rules_distroless/actions/runs/12284400987/job/34281214039

jjmaestro commented:
@thesayyn :S  will look into it now!

jjmaestro commented:
@thesayyn OK, so it's a bit weird, at least to me. I've managed to run all tests in `//examples` manually, e..g going one by one through `bazel test //examples/flatten/...` and so on, except `//examples/debian_snapshot:test` and `//examples/ubuntu_snapshot:test` because they are `SKIPPED` because they are guarded by `target_compatible_with`.

Even so, if I try to run them, I get, as expected
```
% bazel test //examples/debian_snapshot:test --cache_test_results=no
WARNING: Ignoring JAVA_HOME, because it must point to a JDK, not a JRE.
Target //examples/debian_snapshot:test failed to build
Use --verbose_failures to see the command lines of failed build steps.
ERROR: Analysis of target '//examples/debian_snapshot:test' failed; build aborted: Target //examples/debian_snapshot:test is incompatible and cannot be built, but was explicitly requested.
Dependency chain:
    //examples/debian_snapshot:test (efbef2)   <-- target platform (@@platforms//host:host) didn't satisfy constraint @@platforms//os:linux
INFO: Elapsed time: 0.119s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
ERROR: Build did NOT complete successfully
ERROR: No test targets were found, yet testing was requested
```

However! If I run `bazel test //examples/...` then it breaks!!

```
% bazel test //examples/... --cache_test_results=no
WARNING: Ignoring JAVA_HOME, because it must point to a JDK, not a JRE.
ERROR: /private/var/tmp/_bazel_jjmaestro/be9c074a2171a3cb6c673d10113bd570/external/_main~apt~noble/BUILD.bazel:196:12: configurable attribute "controls" in @@_main~apt~noble//:dpkg_status doesn't match this configuration. Would a default condition help?

Conditions checked:
 @@_main~apt~noble//:linux_amd64
 @@_main~apt~noble//:linux_arm64

To see a condition's definition, run: bazel query --output=build <condition label>.

This instance of @@_main~apt~noble//:dpkg_status has configuration identifier 69bb7f0. To inspect its configuration, run: bazel config 69bb7f0.

For more help, see https://bazel.build/docs/configurable-attributes#faq-select-choose-condition.

Use --verbose_failures to see the command lines of failed build steps.
ERROR: Analysis of target '//examples/ubuntu_snapshot:_noble_index_json' failed; build aborted: Analysis failed
INFO: Elapsed time: 0.178s, Critical Path: 0.03s
INFO: 11 processes: 11 internal.
ERROR: Build did NOT complete successfully
//examples/cacerts:test_cacerts                                       NO STATUS
//examples/debian_snapshot:test                                         SKIPPED
//examples/flatten:test_flatten                                       NO STATUS
//examples/flatten:test_flatten_dedup_listing                         NO STATUS
//examples/flatten:test_flatten_dedup_mtree                           NO STATUS
//examples/group:test_group                                           NO STATUS
//examples/group:test_group_content                                   NO STATUS
//examples/home:test_home                                             NO STATUS
//examples/locale:test_bookworm                                       NO STATUS
//examples/locale:test_bullseye                                       NO STATUS
//examples/os_release:test_os_release                                 NO STATUS
//examples/os_release:test_os_release_alternative_path                NO STATUS
//examples/os_release:test_os_release_content                         NO STATUS
//examples/passwd:test_passwd                                         NO STATUS
//examples/passwd:test_passwd_content                                 NO STATUS
//examples/statusd:test_statusd                                       NO STATUS

Executed 0 out of 16 tests: 16 were skipped.
```

So, somehow, something is running on a previous phase / stage? :-? 

Anyway, I'll add a default step to the `select`s but not sure if that will be enough to fix it.

jjmaestro commented:
OK, got it fixed in #136!
